### DISCUSSION TITLE
Buffer size differences vs stock zlib

### DISCUSSION THREAD

**Discussion by Dead2** [zlib-ng/zlib-ng#851] [General]
Due to SSE alignment required by optimizations, zlib-ng will (on affected architectures) require different buffer sizes than zlib itself requires. Some applications such as nginx handles the failure gracefully by logging a warning and working around the issue (likely at a performance penalty), but others just crash randomly without logging any error (php?).

See: http://forum.nginx.org/read.php?2,252113,252133#msg-252133

Now, there are several ways applications can detect forked zlib versions (such as the last part of the version number being .f), and some could be solved through defines in zlib.h.

But perhaps the best solution would be to add a new exported function allowing the actually installed library (might very well differ from the one the application is compiled against) to indicate what its buffer needs are. The application could then cache this value on startup if it wants to avoid running the for each new connection (nginx for example).

Other potential information that the application could need should also be considered at the same time, preferably making something that can be expanded without breaking API/ABI compatability.

**Comment by mtl1979** [comment]
No matter what solution is taken, API is going to change... I rather see the API change in a way that all official releases and forks switch to the new API in consistent way and timeline. At least in Linux world it is pretty easy to detect API conflicts when libraries are installed from packages, but "tarball" and raw installs have to deal with the issue themselves.

**Comment by nemequ** [comment]
Is there any documentation on what the requirements for zlib-ng are?  Perhaps I'm missing something, but all I've seen is that they are different from zlib, no actual numbers…

Also, you mention the size in this issue then go on to talk about alignment.  Is it only alignment that is a concern, or is size really an issue too?

**Comment by RJVB** [comment]
For now I've only meddled with getting this code to build on OS X and configure using cmake, without any real experience in using zlib in my own code. So I wonder:
- does this mean zlib-ng and apparently Intel's zlib version are not actual ABI-compatible drop-in replacements?
- why leave all aspects of buffer allocation to the caller, why not provide at least dedicated allocation and free functions that take care of alignment and (over)sizing? Fftw and FFMpeg do that, IIRC.

**Comment by Dead2** [comment]
Actually the API is pretty clear on this, if you use the initialize functions then it will work just fine. It allocates and frees the required buffers and so on, it even supports using caller-provided malloc/free functions.
Unfortunately some applications rather assume they know the internals of zlib and cheats by allocating a fixed size buffer and does a sloppy initialization instead of calling the initialize function.
Nginx is one example of this.

The reason for zlib-ng, cloudflare-zlib, intel-zlib, etc is using bigger structs is so it can allow SSE/AVX optimizations. This also involves alignment so data is placed in optimal positions directly instead of having to move/copy data around before and after each SSE/AVX instruction call (that would likely eat up most if not all of the performance inprovement).

The struct size differs depending on build options and architecture, have a look in the initalization code to see the ifdef's and so on, it is pretty straight forward. (That having been said, there might be possibilities for optimizations and/or simplification of the code by re-arranging and removing unneccessary variables from the structs.)

So in summary:
Zlib-ng is (should be) drop-in compatible with any well-behaved application.

**Comment by RJVB** [comment]
Ok, good, in that case sorry for the noise :)
(btw, I understood the reasons for the stronger buffer requirements...)

R.

**Comment by maartenbrock** [comment]
If I read this correct, you say that some applications make false assumptions about the memory allocation and take a shortcut. And that it will cost the performance improvement to let the library fix it by moving/copying .

Then why not let those applications suffer the degraded performance (but only those) instead of crashing? Without looking at the code I assume it will mean a single but run time alignment check in every exposed function that accepts a pointer.

**Comment by mtl1979** [comment]
If we add alignment tests, it means the performance degrades for all applications, not just one. We have done this for the PPC version as without them, there would be data corruption for almost all applications.

**Comment by maartenbrock** [comment]
Yes all applications will suffer a small time penalty from the alignment test. But is it really significant when (de)compressing a block? I assume function call overhead is already bigger.

I don't understand your remark about data corruption. When called with unaligned pointers I suggest to call a non-optimized version that doesn't require alignment.

**Comment by mtl1979** [comment]
One extra branch instruction always slow down a lot because it means simple code can't be inlined. Decompression function is called often with short inputs, compression on the otherhand use large buffers with fewer calls.
